contributors/design-proposals/scheduling/scheduling-policy.md

```diff
@@ -0,0 +1,516 @@
+
+_Authors: @arnaudmz, @yastij_
+
+_Reviewers: @bsalamat, @liggitt_
+
+# Objectives
+
+-  Define the concept of scheduling policies
+-  Propose their initial design and scope
+
+## Non-Goals
+
+-  How taints / tolerations work.
+-  How NodeSelector works.
+-  How node / pod affinity / anti-affinity rules work.
+-  How several schedulers can be used within a single cluster.
+-  How priority classes work.
+-  How to set defaults in Kubernetes.
+
+# Background
+
+During real-life Kubernetes architecturing we encountered contexts where role-isolation (between administration and simple namespace usage in a multi-tenant context) could be improved. So far, no restriction is possible on toleration, priority class usage, nodeSelector, anti-affinity depending on user permissions (RBAC).
+
+Identified use-cases aim to ensure that administrators have a way to restrict users or namespaces, it allows administrators to:
+
+-  Restrict execution for specific applications (which are namespace scoped) into certain nodes
+-  Create policies that prevent users from even attempting to schedule workloads onto masters to maximize security
+-  require that a pods under a namespace run on dedicated nodes
+-  Restrict usage of some `PriorityClass`
+-  Restrict usage to a specific set of schedulers.
+-  placing pods on specific nodes (master roles for instance).
+-  using specific priority classes.
+-  expressing pod affinity or anti-affinity rules.
+
+Also Mandatory (and optionally default) values must also be enforced by scheduling policies in case of:
+
+-  mutli-arch (amd64, arm64) of multi-os (Linux, Windows) clusters
+-  multi-az / region / failure domain clusters
+
+# Overview
+
+### syntaxic standards
+
+```yaml
+apiVersion: policy/v1alpha1
+kind: <POLICY_KIND>
+metadata:
+  name: <POLICY_NAME>
+spec:
+  bindingMode: <BINDING_MODE>            # Describes a bindingMode (any or all)
+  namespaces:                            # a list of namespaces that the policy targets (optional)
+    - ns1
+    - ns1
+  namespaceSelector:                     # a selector to match a set of namespaces (optional)
+    key: value
+  rules:                                 # rules that must be satisfied (optional)
+    fieldA:                              # name of the field (optional)
+      - match:                             # Describes how the rule is matched (required)
+        - elt1                           # elements here could be objects like tolerations or strings like SchedulerName
+        action:                            # Describes the action that should be taken (allowed, denied or required)
+
+```
+
+### policy composition
+
+`bindingMode` Describes how policies should be composed:
+
+- any: any policy that has its rules satisfied
+- all: all policies MUST have its rules satisfied
+
+if we have a heterogenous bindingMode across policies (i.e. some policies with any and others with all).
+Then the most restrictive one (i.e the all bindingMode) is applied.
```

Comment:
> most restrictive

How to evaluate and compare the restrictiveness?

Reply:
`all` is more restrictive than `any`

Reply:
It's unclear to me on policy objects level, how the composition happens. E.g. two policy objects: sched_policy_1 and sched_policy_2, both of them define rules for namespace _default_. sched_policy_1 has 5 rules, and sched_policy_2 have 7 rules, how are they evaluated together. The questions are specific to

- Ordering of rules for the same namespace coming from multiple policy objects
- Ordering of rules in different categories (by fields, e.g. schedulerNames vs tolerations vs ...), or if they doesn't matter, how are the results from different categories aggregated together to make a final decision?

Reply:
The algorithm proposed is defined here: https://github.com/kubernetes/community/pull/1937/files#diff-96fed4d55b6bfdccf70f99df676e5cb0R110